### PR TITLE
Tune workflow polling.

### PR DESCRIPTION
**Pull request by gmokki**
- Fetch optimistic locked list of workflow instance ids in separate transaction from the modify phase. This should reduce the risk of deadlocks in databases.
- Also rework code so that if batch returns unsupported values we switch to non-batch usage.

### CODE REVIEW

File: nflow-engine/src/main/java/io/nflow/engine/internal/dao/WorkflowInstanceDao.java [hunk 1]
```diff
@@ -553,69 +552,63 @@ String whereConditionForInstanceUpdate() {
...
+     for (OptimisticLockKey instance : instances) {
+       batchArgs.add(new Object[] { instance.id, sqlVariants.tuneTimestampForDb(instance.modified) });
+       ids.add(instance.id);
+     }
+     int[] updateStatuses = jdbc
+             .batchUpdate(updateInstanceForExecutionQuery() + " where id = ? and modified = ? and executor_id is null", batchArgs);
+     Iterator<Long> idIt = ids.iterator();
+     for (int status : updateStatuses) {
+       idIt.next();
+       if (status == 1) {
```
Comment (efonsell) [inline nflow-engine/src/main/java/io/nflow/engine/internal/dao/WorkflowInstanceDao.java]: I guess it should never happen, but what if status > 1? Now it's handled like status == 1, except that it makes to more checks for the value. So this could be status > 0 here?

File: nflow-engine/src/test/java/io/nflow/engine/internal/dao/WorkflowInstanceDaoTest.java [hunk 2]
```diff
@@ -700,25 +700,6 @@ private WorkflowInstanceDao preparePostgreSQLDao(JdbcTemplate jdbcTemplate) {
     return d;
   }
 
-  @Test
```
Comment (efonsell) [inline nflow-engine/src/test/java/io/nflow/engine/internal/dao/WorkflowInstanceDaoTest.java]: Why is this removed? Is the race condition not happening anymore?

File: nflow-engine/src/test/java/io/nflow/engine/internal/dao/WorkflowInstanceDaoTest.java [hunk 3]
```diff
@@ -1018,7 +999,7 @@ public Poller(WorkflowInstanceDao dao, int batchSize) {
     public void run() {
       try {
         returnSize = min(returnSize, dao.pollNextWorkflowInstanceIds(batchSize).size());
-      } catch (PollingRaceConditionException ex) {
+      } catch (PollingBatchException ex) {
```
Comment (efonsell) [inline nflow-engine/src/test/java/io/nflow/engine/internal/dao/WorkflowInstanceDaoTest.java]: Is this change still valid?

File: nflow-engine/src/main/java/io/nflow/engine/internal/dao/WorkflowInstanceDao.java (hunk 1 above)
Comment (gmokki) [reply nflow-engine/src/main/java/io/nflow/engine/internal/dao/WorkflowInstanceDao.java]: Good point. Rewrote the logic a bit to be shorter and consider values > 1 a failure
